src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,1121 @@
+    if (ec) {
+        _file_size = 0;
+    }
+    _curr_op_progress_callback = progress_callback;
+    fs::path local_path(local_file_path);
+    std::string remote_file_path =
+        remote_folder + std::string("/") + local_path.filename().string();
+    _generic_command_async(CMD_OPEN_FILE_WO, 0, remote_file_path, result_callback);
+}
+
+void MavlinkFTPImpl::_end_write_session()
+{
+    _curr_op = CMD_NONE;
+    if (_ifstream) {
+        _ifstream->close();
+        _ifstream = nullptr;
+    }
+    _terminate_session();
+}
+
+void MavlinkFTPImpl::_write()
+{
+    if (_bytes_transferred >= _file_size) {
+        _session_result = ServerResult::SUCCESS;
+        _end_write_session();
+        return;
+    }
+
+    uint8_t raw_payload[MAVLINK_MSG_FILE_TRANSFER_PROTOCOL_FIELD_PAYLOAD_LEN];
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(raw_payload);
+    payload->seq_number = _seq_number++;
+    payload->session = _session;
+    payload->opcode = _curr_op = CMD_WRITE_FILE;
+    payload->offset = _bytes_transferred;
+    int bytes_read = _ifstream->readsome(reinterpret_cast<char*>(payload->data), max_data_length);
+    if (!*_ifstream) {
+        _end_write_session();
+        _curr_op_result_callback(MavlinkFTP::Result::FILE_IO_ERROR);
+        return;
+    }
+    payload->size = bytes_read;
+    _bytes_transferred += bytes_read;
+    _send_mavlink_ftp_message(raw_payload);
+}
+
+void MavlinkFTPImpl::_terminate_session()
+{
+    if (!_session_valid) {
+        return;
+    }
+    uint8_t raw_payload[MAVLINK_MSG_FILE_TRANSFER_PROTOCOL_FIELD_PAYLOAD_LEN];
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(raw_payload);
+    payload->seq_number = _seq_number++;
+    payload->session = _session;
+    payload->opcode = _curr_op = CMD_TERMINATE_SESSION;
+    payload->offset = 0;
+    payload->size = 0;
+    _send_mavlink_ftp_message(raw_payload);
+}
+
+void MavlinkFTPImpl::list_directory_async(
+    const std::string& path,
+    MavlinkFTP::directory_items_and_result_callback_t callback,
+    uint32_t offset)
+{
+    std::lock_guard<std::mutex> lock(_curr_op_mutex);
+    if (_curr_op != CMD_NONE && offset == 0) {
+        callback(MavlinkFTP::Result::IN_PROGRESS, std::vector<std::string>());
+        return;
+    }
+    if (path.length() >= max_data_length) {
+        callback(MavlinkFTP::Result::INVALID_PARAMETER, std::vector<std::string>());
+        return;
+    }
+
+    _last_path = path;
+    _curr_dir_items_result_callback = callback;
+    _list_directory(offset);
+}
+
+void MavlinkFTPImpl::_list_directory(uint32_t offset)
+{
+    uint8_t raw_payload[MAVLINK_MSG_FILE_TRANSFER_PROTOCOL_FIELD_PAYLOAD_LEN];
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(raw_payload);
+    payload->seq_number = _seq_number++;
+    payload->session = 0;
+    payload->opcode = _curr_op = CMD_LIST_DIRECTORY;
+    payload->offset = offset;
+    strcpy(reinterpret_cast<char*>(payload->data), _last_path.c_str());
```

Comment:
I'm scared. Can we always use `strncpy`?

Reply:
Agree!